CAN bus implementation

bclbruno opened the pull request:
- Add CAN hardware implementation
  - Added CAN bus configuration to HardwarePage: TxPin, RxPin, Baudrate, NodeId
  - Added CAN bus configuration to Settings structure
  - Added CAN bus pins to markup and string generation
  - Added sendcan and sendtocan commands
- [C20] Add CAN controller
- [P155] Add CAN importer plugin

TODOs:
 - Set the plugin and controller IDs (It's set as P155 and C20 just as a place holder)
 - doc?

bruno-clima committed:
CAN Hardware Implementation

bruno-clima committed:
CAN controller implementation

bruno-clima committed:
Plugin implementation

bruno-clima committed:
Fix Markup

tonhuisman commented:
@bclbruno I have assigned Plugin ID `P174` for the CAN Importer plugin and Controller ID `C022` for the CAN Controller, as these are the next available IDs. Please update your code, as the current IDs are already in use.

tonhuisman commented on src/_P155_CAN_helper.ino:
```diff
@@ -0,0 +1,231 @@
+        String lines[valueCount];
+        LoadCustomTaskSettings(event->TaskIndex, lines, valueCount, 0);
+
+        for (int i = 0; i < valueCount; ++i) {
+        if (lines[i].length() == 0) {
+          lines[i] = concat(F("val"), i + 1);
+        }
+          ExtraTaskSettings.setTaskDeviceValueName(i, lines[i]);
+        }
```
Please use Uncrustify to format the sources (exception: external libraries), to avoid mis-alignments like these.

tonhuisman commented on src/_P155_CAN_helper.ino:
```diff
@@ -0,0 +1,231 @@
+    if (!(PCONFIG(3) & idx_flag))
+    {
+      String log_debug = F("Disconnected");
+      addLogMove(LOG_LEVEL_INFO, log_debug);
```
To preserve code-size and memory use, this can be a single statement, no need to instantiate that `String` first.
`addLog(LOG_LEVEL_INFO, F("Disconnected"));`

tonhuisman commented on src/src/Commands/InternalCommands.cpp:
```diff
@@ -1,488 +1,494 @@
+    case ESPEasy_cmd_e::wifissid:                   COMMAND_CASE_R(Command_Wifi_SSID,       1);      // WiFi.h
+    case ESPEasy_cmd_e::wifissid2:                  COMMAND_CASE_R(Command_Wifi_SSID2,      1);      // WiFi.h
+    case ESPEasy_cmd_e::wifistamode:                COMMAND_CASE_R(Command_Wifi_STAMode,    0);      // WiFi.h
+
+    case ESPEasy_cmd_e::sendcan:                    COMMAND_CASE_A(Command_CAN_SendCAN,    -1);      //CAN.h
+    case ESPEasy_cmd_e::sendtocan:                  COMMAND_CASE_A(Command_CAN_SendToCAN,  -1);      //CAN.h
+
```
- There's a missing `#if FEATURE_CAN` here
- Commands are in this switch statement in alphabetical order, by default

tonhuisman commented on src/src/Commands/InternalCommands_decoder.cpp:
```diff
@@ -276,6 +276,10 @@ const char Internal_commands_w[] PROGMEM =
   "wdconfig|"
   "wdread|"
 #endif // ifndef LIMIT_BUILD_SIZE
+#ifdef FEUTER_CAN
```
- That's probably a typo, should better be `#if FEATURE_CAN`
- We've stopped using `#ifdef` for feature-flags, to make handling defaults possible, that can also be controller from a `Custom.h` file, to be explicitly enabled (1) or disabled (0)

tonhuisman commented on src/src/CustomBuild/define_plugin_sets.h:
```diff
@@ -1452,6 +1452,11 @@ To create/register a plugin, you have to :
     #define USES_C013   // ESPEasy P2P network
 #endif
 
+#ifdef FEATURE_CAN
+    #define USES_C020 //CAN
+    #define USES_C021
+    #define USES_P155
+#endif
```
This should be like:
```
#ifndef FEATURE_CAN
  #ifdef ESP32
    #define FEATURE_CAN 1
  #endif
  #ifdef ESP8266
    #define FEATURE_CAN 0 // Default disabled for ESP8266 for size reasons
  #endif
#endif
#if FEATURE_CAN
  #ifndef USES_C022 // CAN Controller
    #define USE_C022
  #endif
  #ifndef USES_P174
    #define USES_P174 // CAN Import plugin
  #endif
#endif
```
(For Plugin, Controller and Notifiers we still use `#ifdef USES_...` checks...)
Not sure why you also try to enable Controller C021? If that's one you also created yourself, then it should probably get ID C023, as C021 is reserved for a (planned) LoRa controller 🤔

tonhuisman commented on src/src/DataStructs/SettingsStruct.h:
```diff
@@ -567,6 +570,11 @@ class SettingsStruct_tmpl
   int8_t        console_serial_txpin = DEFAULT_CONSOLE_PORT_TXPIN;
   uint8_t       console_serial0_fallback = DEFAULT_CONSOLE_SER0_FALLBACK;
+
+  uint8_t CAN_Tx_pin   = -1;
+  uint8_t CAN_Rx_pin   = -1;
+  long    CAN_baudrate = DEFAULT_CAN_BAUDRATE;
+  int     CAN_node_id  = 0;
```
I think this belongs in the CAN Controller, where the CAN Importer can get it's pins from, if needed, as the CAN Controller seems t o be a requirement for the importer to function.

TD-er committed:
Merge branch 'mega' into can_implementation

TD-er committed:
[CAN] Make it compile

TD-er committed:
[CAN] Rename plugin and controller to reserved IDs

TD-er committed:
[CAN] Fix command decoder for CAN commands

TD-er committed:
[CAN] Make lib compile for ESP-IDF5.1

TD-er committed:
[CAN] Limit to ESP32-classic only

Current implementation can only build on ESP32-classic. May need significant code changes for ESP32C3/C6/S2/S3

TD-er committed:
[CAN] custom_IR_ESP32c3_4M316k_CDC did not ignore CAN lib

TD-er committed:
[CAN] Some more envs with their own lib_ignore not ignoring CAN lib

TD-er committed:
[CAN] FIX ESP8266 builds

TD-er committed:
[CAN] Fix numerous other ESP8266 builds lib_ignore CAN library

tonhuisman commented on src/src/CustomBuild/define_plugin_sets.h:
```diff
@@ -3438,7 +3463,11 @@ To create/register a plugin, you have to :
 
 
-  
+#if FEATURE_CAN
+  #define USES_C022 //CAN
+  #define USES_P174
+#endif
```
This is a duplicate for lines 3431..3441

TD-er committed:
[CAN] Fix settings struct size check

TD-er committed:
[CAN] And another few envs needing CAN lib_ignore

TD-er committed:
Merge branch 'mega' into can_implementation

TD-er committed:
Merge branch 'mega' into can_implementation

TD-er committed:
Merge branch 'mega' into can_implementation

TD-er committed:
Merge branch 'mega' into can_implementation

TD-er committed:
Merge branch 'mega' into can_implementation

TD-er committed:
Merge remote-tracking branch 'letscontrolit/mega' into can_implementation

TD-er committed:
Merge branch 'mega' into can_implementation

tonhuisman commented on src/src/Helpers/_Plugin_Helper_CAN.cpp:
```diff
@@ -0,0 +1,214 @@
+  addLogMove(LOG_LEVEL_INFO, log);
+
+  for (taskIndex_t x = 0; x < TASKS_MAX; x++) {
+    constexpr pluginID_t PLUGIN_ID_CAN_HELPER(155);
```
That 155 seems to be a typo, AFAICS the helper plugin (now) has PluginID 174.

tonhuisman commented on src/src/Helpers/_Plugin_Helper_CAN.cpp:
```diff
@@ -0,0 +1,214 @@
+    constexpr pluginID_t PLUGIN_ID_CAN_HELPER(155);
+    if (Settings.TaskDeviceEnabled[x] 
+        && (Settings.getPluginID_for_task(x) == PLUGIN_ID_CAN_HELPER)
+        && (Settings.TaskDevicePluginConfig[x][0] == tmp.Par1)
+        && (Settings.TaskDevicePluginConfig[x][2] == tmp.Par2)
```
Please use the `PCONFIG()` macro.

TD-er committed:
Merge branch 'mega' into can_implementation

tonhuisman commented on src/_P174_CAN_helper.ino:
```diff
@@ -0,0 +1,233 @@
+    {
+      String lines[VARS_PER_TASK];
+
+      addRowLabel(F("Node"));
```
For the `PLUGIN_WEBFORM_SAVE` function, no new UI elements should be added.

tonhuisman commented on src/_P174_CAN_helper.ino:
```diff
@@ -0,0 +1,233 @@
+    case PLUGIN_READ:
+    {
+
+      addLogMove(LOG_LEVEL_DEBUG, strformat(
```
`DEBUG` level logging must be wrapped with `#ifndef BUILD_NO_DEBUG` / `#endif`

tonhuisman commented on src/_P174_CAN_helper.ino:
```diff
@@ -0,0 +1,233 @@
+
+      for (int i = 0; i < VARS_PER_TASK; ++i) {
+        lines[i] = webArg(concat(F("p174_value"), i));
+        if (lines[i].length() == 0) {
```
The String class has an `isEmpty()` method available (multiple occurrences found)

TD-er committed:
Merge branch 'mega' into can_implementation

TD-er committed:
Merge remote-tracking branch 'letscontrolit/mega' into can_implementation

TD-er committed:
Merge branch 'mega' into can_implementation

TD-er committed:
Merge branch 'mega' into can_implementation

TD-er committed:
Merge branch 'mega' into can_implementation

TD-er committed:
Merge branch 'mega' into can_implementation

TD-er committed:
Merge remote-tracking branch 'letscontrolit/mega' into can_implementation

TD-er committed:
Merge branch 'mega' into can_implementation

TD-er commented:
There are quite a few merge conflicts.
Can you please have a look at this?

TD-er committed:
Merge branch 'mega' into can_implementation

TD-er commented:
I tried to merge it as it also required some manual interactions. Hopefully I didn't break anything :)

TD-er committed:
[CAN] Fix merge issues

TD-er committed:
[CAN] Fixed failing ESP8266 build

TD-er committed:
[CAN] Fix failing build on ESP32S2

TD-er committed:
[CAN] Fix failing custom_IR_ESP32s2_4M316k

TD-er committed:
[CAN] Fix failing build for custom_IR_ESP32s3_4M316k

TD-er committed:
[CAN] Fixed failing max_ESP32c6 builds

TD-er commented:
Hopefully the build will now complete successfully.
There are more ESP32-variants supporting CAN/TWAI so maybe it is a good idea to look for using either another lib or draw inspiration from a library like this one: https://github.com/handmade0octopus/ESP32-TWAI-CAN/blob/master/src/ESP32-TWAI-CAN.cpp

There are also examples (now, likely not when you started the PR) for using TWAI in arduino-esp32: https://github.com/espressif/arduino-esp32/tree/master/libraries/ESP32/examples/TWAI

This way it doesn't depend on such highly platform specific code as used in the lib currently included in the PR.

I did have to move the settings members to the end of the settings struct, as there have been some network related settings added to the Settings struct and the original location would have rendered existing settings without this PR code incompatible.

Right now the code is up-to-date with the current `mega` branch.
